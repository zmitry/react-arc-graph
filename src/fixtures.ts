export const data1 = [
  {
    from: "1",
    to: "2"
  },
  {
    from: "1",
    to: "3"
  },

  {
    from: "1",
    to: "4"
  },
  {
    from: "2",
    to: "4"
  },
  {
    from: "4",
    to: "5"
  }
];
export const data4 = [
  {
    from: "1",
    to: "2"
  },
  {
    from: "1",
    to: "3"
  },

  {
    from: "2",
    to: "4"
  },
  {
    from: "3",
    to: "4"
  }
];
export const data2 = [
  {
    from: 1,
    to: 3
  },
  {
    from: 2,
    to: 3
  },
  {
    from: 3,
    to: 8
  },
  {
    from: 4,
    to: 6
  },
  {
    from: 4,
    to: 6
  },
  {
    from: 4,
    to: 7
  },
  {
    from: 7,
    to: 8
  },
  {
    from: 8,
    to: 9
  },
  {
    from: 4,
    to: 9
  },

  {
    from: 9,
    to: 10
  },
  {
    from: 9,
    to: 11
  },
  {
    from: 11,
    to: 12
  },
  {
    from: 11,
    to: 13
  },
  {
    from: 11,
    to: 14
  },
  {
    from: 11,
    to: 15
  },
  {
    from: 11,
    to: 16
  },
  {
    from: 12,
    to: 17
  },
  {
    from: 13,
    to: 19
  },
  {
    from: 2,
    to: 5
  },

  {
    from: 3,
    to: 7
  },
  {
    from: 3,
    to: 8
  },
  {
    from: 7,
    to: 8
  },
  {
    from: 5,
    to: 8
  },
  {
    from: 6,
    to: 8
  },
  {
    from: 1,
    to: 9
  },
  {
    from: 2,
    to: 9
  },
  {
    from: 1,
    to: 8
  }
];

export const paddingMockData = [
  {
    fromIndex: 0,
    index: 0,
    toIndex: 1,
    side: 1,
    pad: 0
  },
  {
    fromIndex: 0,
    index: 1,
    toIndex: 2,
    side: 1,
    pad: 1
  },

  {
    index: 2,
    fromIndex: 1,
    toIndex: 3,
    side: 1,
    pad: 2
  },
  {
    index: 3,
    fromIndex: 2,
    toIndex: 4,
    side: 1,
    pad: 1
  }
];
export const paddingMockData2 = [
  {
    fromIndex: 0,
    index: 0,
    toIndex: 0,
    side: 1,
    pad: 0
  }
];
