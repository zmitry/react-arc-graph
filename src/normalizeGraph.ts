import { range } from 'lodash';

import { IEdgeInternal } from './ListViewGraph';

export function getEdgeSides(edges: IEdgeInternal[]) {
  // make kv map for edges value is nodes indexes and values in array of edges which cross this coordinates
  // for e.g we have something like this
  // where short nodes are 1 and to respectively and long node is 3
  // and nodes are 1 2 3 respectively
  // * - * - *
  //  \_____/
  // in result we will have map like this
  // 1 -> [1,3]
  // 2 -> [2,3]
  const kvMap = edges.reduce<Map<number, IEdgeInternal[]>>((acc, el) => {
    if (Math.abs(el.toIndex - el.fromIndex) === 1) {
      return acc;
    }
    for (let i = el.fromIndex; i < el.toIndex; i++) {
      const v = acc.get(i) || [];
      acc.set(i, [...v, el]);
    }
    return acc;
  }, new Map());

  // map where key is index of edge in array and value is side which you need to set
  const itemsToSet = [...kvMap.entries()]
    .sort((a, b) => a[0] - b[0])
    .reduce<Map<number, number>>((acc, [__, lineEdges]) => {
      // compute edges which are already on the left side
      const leftEdges = lineEdges.filter(el => acc.get(el.index) === -1);
      let diff = Math.ceil((lineEdges.length - leftEdges.length) / 2);
      for (const edge of lineEdges) {
        const index = edge.index;
        if (acc.get(index) === undefined) {
          // set left direction until it's half of all edges on this space
          const side = edge.index % 2 === 1 && diff > 0 ? -1 : 1;
          acc.set(index, side);

          if (side === -1) {
            diff--;
          }
        }
      }
      return acc;
    }, new Map());

  return itemsToSet;
}

export interface IPaddingFinderOptions {
  defaultPad?: number;
  padStep?: number;
  allowOneSepIntersections?: boolean;
}

export const defaultPaddingFinderOptions: IPaddingFinderOptions = {
  defaultPad: 1,
  padStep: 1,
  // if true we will allow to be element like this on the same lvl
  // * - * - * - *
  //  \___\_/___/
  // if false it will be like this
  // * - * - * - *
  //  \___\_/   /
  //       \___/
  allowOneSepIntersections: false,
};

// computes padding between graph center and edge
export function getEdgePaddings(
  edges: IEdgeInternal[],
  options = defaultPaddingFinderOptions
) {
  const { defaultPad, padStep, allowOneSepIntersections } = Object.assign(
    {},
    defaultPaddingFinderOptions,
    options
  );

  const kvMap = new Map<string, number>();
  // dont care about 1 length edges
  const filtered = edges
    .filter(el => el.pad !== 0)
    .sort((a, b) => a.pad - b.pad);
  // key is edge index and value is padding to set
  const result = new Map<number, number>();
  filtered.forEach(el => {
    // find max padding through all intersected edges
    /* tslint:disable */
    let [a, b] = [el.fromIndex, el.toIndex].sort((a, b) => a - b);
    /* tslint:enable */
    if (allowOneSepIntersections) {
      a += 1;
    }
    const ranges = range(a, b).map(i => {
      const k = i + ':' + el.side;
      return kvMap.get(k) || defaultPad;
    });
    const maxPad = ranges.length > 0 ? Math.max(...ranges) : defaultPad;

    const nextPad = maxPad + padStep;
    let startIndex = el.fromIndex;
    let endIndex = el.toIndex;
    if (allowOneSepIntersections) {
      startIndex += 1;
      endIndex -= 1;
    }

    // delete that padding from all intersected
    for (let i = startIndex; i <= endIndex; i++) {
      const k = i + ':' + el.side;
      kvMap.set(k, nextPad);
    }
    result.set(el.index, nextPad);
  });
  return result;
}
