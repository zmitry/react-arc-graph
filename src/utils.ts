export const nodesFromEdges = (d: any[]) => {
  return [...new Set(d.flatMap(el => [el.from, el.to]))]
    .sort()
    .map(id => ({ id: id }));
};
