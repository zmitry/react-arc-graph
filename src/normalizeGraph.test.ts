import { data1, data2, paddingMockData, paddingMockData2 } from './fixtures';
import { getEdgePaddings, getEdgeSides } from './normalizeGraph';
import { nodesFromEdges } from './utils';
const toNodePrefix = '__to';
const fromNodePrefix = '__from';

export const createFakeFromToNodes = (nodes: any[]): any[] => {
  return nodes.reduce(
    (acc, node) =>
      acc.concat([
        { ...node, fakeId: node.id + toNodePrefix },
        { ...node, fakeId: node.id + fromNodePrefix },
      ]),
    []
  );
};

const mapEdges = edges => edges;

export const createFakeNodesIds = nodes => nodes.map(node => node.fakeId);

const createNodesIds = edges =>
  createFakeNodesIds(createFakeFromToNodes(nodesFromEdges(edges)));

describe('ArcGraph', () => {
  it('finds sides', () => {
    const edges = mapEdges({
      edges: data1,
      nodeIDs: createNodesIds(data1),
      defaultSide: 'auto',
    });

    const mapping = getEdgeSides(edges);
    expect(mapping).toMatchSnapshot();
    const edges2 = mapEdges({
      edges: data2,
      nodeIDs: createNodesIds(data2),
      defaultSide: 'auto',
    });
    const mapping2 = getEdgeSides(edges2);
    expect(mapping2).toMatchSnapshot();
  });

  it('it finds padding ', () => {
    const mapping = getEdgePaddings(paddingMockData as any, {
      defaultPad: 0,
      padStep: 1,
    });
    expect(mapping).toMatchInlineSnapshot(`
Map {
  1 => 1,
  3 => 2,
  2 => 3,
}
`);
    const mapping2 = getEdgePaddings(paddingMockData as any, {
      allowOneSepIntersections: true,
      defaultPad: 0,
      padStep: 1,
    });

    expect(mapping2).toMatchInlineSnapshot(`
Map {
  1 => 1,
  3 => 1,
  2 => 1,
}
`);

    const mapping3 = getEdgePaddings(paddingMockData2 as any, {
      allowOneSepIntersections: true,
      defaultPad: 0,
      padStep: 1,
    });
    expect(mapping3.size).toBe(0);
  });
});
