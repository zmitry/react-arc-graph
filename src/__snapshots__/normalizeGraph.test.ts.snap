// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArcGraph finds sides 1`] = `
Map {
  1 => -1,
  2 => 1,
  3 => -1,
}
`;

exports[`ArcGraph finds sides 2`] = `
Map {
  0 => 1,
  24 => 1,
  26 => 1,
  9 => -1,
  10 => 1,
  12 => 1,
  13 => -1,
  14 => 1,
  15 => -1,
  16 => 1,
  17 => -1,
  18 => 1,
  25 => -1,
  2 => 1,
  19 => -1,
  20 => 1,
  3 => -1,
  4 => 1,
  5 => -1,
  8 => 1,
  22 => 1,
  23 => -1,
}
`;

exports[`ArcGraph mapEdges 1`] = `
Array [
  Object {
    "direction": -1,
    "from": "1",
    "fromIndex": 1,
    "index": 0,
    "pad": 0,
    "side": 1,
    "to": "2",
    "toIndex": 2,
  },
  Object {
    "direction": -1,
    "from": "1",
    "fromIndex": 1,
    "index": 1,
    "pad": 2,
    "side": 1,
    "to": "3",
    "toIndex": 4,
  },
  Object {
    "direction": -1,
    "from": "1",
    "fromIndex": 1,
    "index": 2,
    "pad": 4,
    "side": 1,
    "to": "4",
    "toIndex": 6,
  },
  Object {
    "direction": -1,
    "from": "2",
    "fromIndex": 3,
    "index": 3,
    "pad": 2,
    "side": 1,
    "to": "4",
    "toIndex": 6,
  },
  Object {
    "direction": -1,
    "from": "4",
    "fromIndex": 7,
    "index": 4,
    "pad": 0,
    "side": 1,
    "to": "5",
    "toIndex": 8,
  },
]
`;
